instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/SpanKey.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.instrumenter;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.ContextKey;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+// TODO make class (and fields/methods below) package-private after tracers are gone and no need for
+//  bridging from ClientSpan/ServerSpan/ConsumerSpan
+public final class SpanKey {
+
+  private static final ContextKey<Span> SERVER_KEY =
+      ContextKey.named("opentelemetry-traces-span-key-server");
+  private static final ContextKey<Span> CONSUMER_KEY =
+      ContextKey.named("opentelemetry-traces-span-key-consumer");
+
+  // TODO bridge these constants in AgentContextStorage
+  private static final ContextKey<Span> HTTP_KEY =
+      ContextKey.named("opentelemetry-traces-span-key-http");
+  private static final ContextKey<Span> RPC_KEY =
+      ContextKey.named("opentelemetry-traces-span-key-rpc");
+  private static final ContextKey<Span> DB_KEY =
+      ContextKey.named("opentelemetry-traces-span-key-db");
+  private static final ContextKey<Span> MESSAGING_KEY =
+      ContextKey.named("opentelemetry-traces-span-key-messaging");
+
+  // this is used instead of above, depending on the configuration value for
+  // otel.instrumentation.experimental.span-suppression-by-type
+  // named outgoing because it covers both client and producer spans
```

Comment:
Previously we didn't suppress PRODUCER spans at all - should we maintain that behavior? Having a PRODUCER -> CLIENT relationship happens a few times in our messaging instrumentations (a producer calls a http/rpc client that actually sends the message to the broker)

Reply:
I guess we never suppressed them because there as never 2 producer spans so far and if there were, we'd want to suppress nested ones? With suppression-by-type we'd not suppress HTTP client under messaging producer because types are different. You're right that without it we'd start suppressing them.

BTW, it seems messaging producer [can sometimes be client](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md), not sure if we follow it though.

> A producer of a message should set the span kind to PRODUCER unless it synchronously waits for a response: then it should use CLIENT. 

Anyway, I've separated producer strategy from client and we'd suppress nested producers, but never suppress CLIENT under PRODUCER.

Reply:
> I guess we never suppressed them because there as never 2 producer spans so far and if there were, we'd want to suppress nested ones?

I that case we probably would want to suppress them. 

> BTW, it seems messaging producer [can sometimes be client](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md), not sure if we follow it though.

🤯 
I don't think that we follow it; we always use `PRODUCER` for all outgoing messaging spans. (Anyway, that's another reason why not using the span kind when suppressing might be a good idea)